handlers/http.go

```diff
@@ -46,12 +51,17 @@ func (hrl httpRateLimiter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	matches, err := hrl.rateLimiter.Add(request)
 	if err != nil && err != leakybucket.ErrorFull {
 		log.Printf("[%s] ERROR: %s", guid, err)
-		w.WriteHeader(500)
-		return
+		if !hrl.AllowOnError {
+			log.Printf("[%s] WARNING: bypassing rate limiter due to Error")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
 	}
+
 	addRateLimitHeaders(w, matches)
```

Comment:
@MaerF0x0 and I talked through this a bit offline, but I found it confusing that this call to `addRateLimitHeaders` doesn't get skipped when `AllowOnError` is `true` and we have an error. I think this could benefit from a comment to clarify that we expect `matches` to be empty in that case or a restructuring to something like this:

```go
switch {
case err == leakybucket.ErrorFull:
  addRateLimitHeaders(w, matches)
  w.WriteHeader(StatusTooManyRequests)
case err != nil && hrl.AllowOnError {
  hrl.proxy.ServeHTTP(w, r)
case err != nil {
  w.WriteHeader(http.StatusInternalServerError)
default:
  addRateLimitHeaders(w, matches)
  hrl.proxy.ServeHTTP(w, r)
}
```